src/content/en/2020/http2.md

```diff
@@ -4,20 +4,534 @@ part_number: IV
+
+## Getting to a better protocol
+
+Let’s say a client and server support both HTTP/1.1 and HTTP/2. How do they choose which one to use? The most common method is the TLS Application Layer Protocol Negotiation (ALPN). Using this, clients send a list of protocols they support to the server, which picks the one it prefers to use for that connection.  Because the browser needs to negotiate the TLS parameters as part of setting up an HTTPS connection, it can also negotiate the HTTP version at the same time. Since both HTTP/2 and HTTP/1.1 can be served from the same TCP port (443), browsers don’t need to make this selection before opening a connection.
+
+This doesn’t work if the protocols aren’t on the same port, use a different transport protocol (TCP vs QUIC), or if you’re not using TLS. For those scenarios, you start with whatever is available on the first port you connect to, then discover other options.  HTTP defines two mechanisms to change protocols for an origin after connecting.
+
+### Upgrade
+
+The Upgrade header has been part of HTTP for a long time.  In HTTP/1.x, Upgrade allows a client to make a request using one protocol, but indicate its support for another protocol (like HTTP/2).  If the server also supports the offered protocol, it responds with a status 101 (Switching Protocols) and proceeds to answer the request in the new protocol.  If not, the server answers the request in HTTP/1.x.  Servers can advertise their support of a different protocol using an Upgrade header on a response.
+
+The most common application of Upgrade is Websockets.  HTTP/2 also defines an Upgrade path, for use with its unencrypted cleartext mode.  There is no support for this capability in web browsers however.  Therefore, it’s not surprising that less than 3% of HTTP/1.1 requests in our dataset received an Upgrade header in the response.  A very small number (0.0011% of HTTP/2, 0.064% of HTTP/1.1) of requests using TLS also received Upgrade headers in response; these are likely cleartext HTTP/1.1 servers behind intermediaries which speak HTTP/2 and/or terminate TLS, but don’t properly remove Upgrade headers.
+
+### Alternative Services
+
+Alternative Services (Alt-Svc) enables an HTTP origin to indicate other endpoints which serve the same content, possibly over different protocols.  Although uncommon, HTTP/2 might be located at a different port or different host from a site’s HTTP/1.1 service.  More importantly, since HTTP/3 uses UDP where prior versions of HTTP use TCP, HTTP/3 will always be at a different endpoint from the HTTP/1.x and HTTP/2 service.
```

Comment:
```suggestion
Alternative Services (Alt-Svc) enables an HTTP origin to indicate other endpoints which serve the same content, possibly over different protocols.  Although uncommon, HTTP/2 might be located at a different port or different host from a site’s HTTP/1.1 service.  More importantly, since HTTP/3 uses QUIC (hence UDP) where prior versions of HTTP use TCP, HTTP/3 will always be at a different endpoint from the HTTP/1.x and HTTP/2 service.
```